### ISSUE TITLE
flush.console() does not seem to work 

### ISSUE THREAD

**Issue by gaborcsardi** [rstudio/rstudio#8040]
### System details

    RStudio Edition : desktop
    RStudio Version : 1.4.867
    OS Version      : macOS 10.14.6
    R Version       : 4.0.2

### Steps to reproduce the problem

RUn this:

```r
fun <- function() { 
  cat("select the file in the dialog box\n")
  flush.console()
  file.choose()
}
fun()
```

### Describe the problem in detail

`flush.console()` is supposed to flush the console buffer, so the in the code above the `cat()`-ed text should appear, before the `file.choose()` dialog appears.

### Describe the behavior you expected

I expected the `cat()`-ed text to appear in the console before the file chooser dialog.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting a RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by kevinushey** [comment]
Reproduces for me:

<img width="868" alt="Screen Shot 2020-10-13 at 9 40 56 AM" src="https://user-images.githubusercontent.com/1976582/95890122-339a1e00-0d38-11eb-9e38-c84ac8accbf8.png">

Note `fun()` called in the console but no other console output available.

That said, we explicitly don't hook into `R_FlushConsole()` and friends:

https://github.com/rstudio/rstudio/blob/06503e51e2df2d183e055454eb8b51b8250a3ff6/src/cpp/r/session/REmbeddedPosix.cpp#L151-L155

but it seems like it would be reasonable to use those so that we could synchronize on any pending console text that needs to be rendered in the browser.

All that said... definitely not for 1.4! An appropriate workaround here would be something like:

```
if (identical(.Platform$GUI, "RStudio"))
      Sys.sleep(0.1)
```

after calling `flush.console()`.

**Comment by gaborcsardi** [comment]
Does `Sys.sleep(0.1)` ensure that the output is printed? How low can you go?

**Comment by kevinushey** [comment]
Good question -- I'm not sure, and this is definitely a hack that might fail 0.1% of the time :-)

**Comment by gaborcsardi** [comment]
Yeah, it is not really worth it then, I would say. It would be great to fix, if not in 1.4, then later, because RGui and R.app both support this.

**Comment by gaborcsardi** [comment]
This is still an issue, please don't close it.

**Comment by eragrostis** [comment]
It would be great to have some movement on this. I'm rather frustrated running long loops and having no idea what progress my code has made as I cannot print to the console from within a loop, even when using flush.console()!

**Comment by kevinushey** [comment]
Does using `Sys.sleep(0.1)` instead of `flush.console()` function as a workaround?

**Comment by eragrostis** [comment]
I apologize, In the process of creating a working example and it turns out there's another bug behind this. It's that I'm running ggplot2::ggsave() inside my loop which is silencing any message()'s run in the loop after ggsave() is called. This issue also only appears in R markdown. Using cat() appears to be a solution for this problem. 

Here is a reproducible example: 

```{r}
library(ggplot2)
for(i in 1:3){
  message("message") #appears once in console, expect 3 appearances
  plot <- ggplot()
  ggsave("junk.png") #comment this out and you see the expected console output
  
  message("message2") #does not appear in console
  cat("\nfinish") #appears 3 times as expected
}
message(message3) #appears once as expected
```

Should I create a new issue on the R markdown project?

**Comment by kevinushey** [comment]
I think this still requires further investigation; it's not clear to me if this is an issue on the RStudio side or not.
